api/GrpcServer.go

```diff
@@ -0,0 +1,767 @@
+}
+
+// UpdateRepo updates GitMaterial
+func (controller *GrpcControllerImpl) UpdateRepo(ctx context.Context, req *pb.GitMaterial) (
+	*pb.Empty, error) {
+
+	// Mapping
+	mappedGitMaterial := &sql.GitMaterial{
+		Id:               int(req.Id),
+		GitProviderId:    int(req.GitProviderId),
+		Url:              req.Url,
+		FetchSubmodules:  req.FetchSubmodules,
+		Name:             req.Name,
+		CheckoutLocation: req.CheckoutLocation,
+		CheckoutStatus:   req.CheckoutStatus,
+		CheckoutMsgAny:   req.CheckoutMsgAny,
+		Deleted:          req.Deleted,
+	}
+
+	// Update repo
+	_, err := controller.repositoryManager.UpdateRepo(mappedGitMaterial)
+	if err != nil {
+		controller.logger.Errorw("error while updating repo",
+			"name", mappedGitMaterial.Name,
+			"err", err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return &pb.Empty{}, nil
+}
+
+// SavePipelineMaterial saves pipeline material
+func (controller *GrpcControllerImpl) SavePipelineMaterial(ctx context.Context, req *pb.SavePipelineMaterialRequest) (
+	*pb.Empty, error) {
+
+	// Mapping to sql package specified struct type
+	ciPipelineMaterials := make([]*sql.CiPipelineMaterial, 0)
+	for _, item := range req.CiPipelineMaterials {
+
+		ciPipelineMaterials = append(ciPipelineMaterials, &sql.CiPipelineMaterial{
+			Id:            int(item.Id),
+			GitMaterialId: int(item.GitMaterialId),
+			Type:          sql.SourceType(item.Type),
+			Value:         item.Value,
+			Active:        item.Active,
+			LastSeenHash:  item.LastSeenHash,
+			CommitAuthor:  item.CommitAuthor,
+			CommitDate:    item.CommitDate.AsTime(),
+			CommitHistory: item.CommitHistory,
+			Errored:       item.Errored,
+			ErrorMsg:      item.ErrorMsg,
+		})
+	}
+
+	// TODO: Check if we can change the argument type for the below method to avoid mapping
+	_, err := controller.repositoryManager.SavePipelineMaterial(ciPipelineMaterials)
+	if err != nil {
+		controller.logger.Errorw("error while adding repo",
+			"err", err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return &pb.Empty{}, nil
+}
+
+// FetchChanges
+func (controller *GrpcControllerImpl) FetchChanges(ctx context.Context, req *pb.FetchScmChangesRequest) (
+	*pb.MaterialChangeResponse, error) {
+
+	res, err := controller.repositoryManager.FetchChanges(int(req.PipelineMaterialId), req.From, req.To, int(req.Count))
+	if err != nil {
+		controller.logger.Errorw("error while fetching scm changes",
+			"pipelineMaterialId", req.PipelineMaterialId,
+			"err", err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if res == nil {
+		return nil, nil
+	}
+
+	// Mapping GitCommit
+	pbGitCommits := make([]*pb.GitCommit, 0)
+	for _, item := range res.Commits {
+
+		mappedCommit, err := controller.mapGitCommit(item)
+		if err != nil {
+			continue
+		}
+		pbGitCommits = append(pbGitCommits, mappedCommit)
+	}
+
+	return &pb.MaterialChangeResponse{
+		Commits:        pbGitCommits,
+		LastFetchTime:  timestamppb.New(res.LastFetchTime),
```

Comment:
have to check for zero value here too. I also think we could write a wrapper for timestamppb.New() which checks for zero values and returns result accordingly since timestamppb.New() gives corrupt result with zero value this may cause problems if some other developer uses it directly.

Reply:
resolved